oadp-1.3: OADP-4265 Mark InProgress backup/restore as failed upon requeuing

kaovilai opened the pull request:
Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

kaovilai committed:
Mark InProgress backup/restore as failed upon requeuing

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

remove uuid, return err to requeue instead of requeue: true

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

kaovilai commented:
from prior feedback need to do finalizer controller also

kaovilai commented:
> from prior feedback need to do finalizer (operations in 1.3) controller also

For restore operations controller we're covered by this returning error: https://github.com/openshift/velero/blob/c06b13825cd7d2df1ea3b70d0df725006127d92e/pkg/controller/restore_operations_controller.go#L194

backup operations controller: https://github.com/openshift/velero/blob/c06b13825cd7d2df1ea3b70d0df725006127d92e/pkg/controller/backup_operations_controller.go#L207

sseago commented:
Yes, I think we're fine leaving the backup/restore_operations_controller returns alone, since those already requeue. I don't think we have restore finalizer in 1.3, so we just need backup, restore, and backup finalizer.

kaovilai commented:
If patch fails on finalizer controller we wanna retry patch as fail? It's kinda difficult there since it's currently using defer func.

I would have to break the patch call out of defer to return reconciler err on patch fail. Is that ok with you?

sseago commented on pkg/controller/backup_controller.go:
```diff
@@ -233,6 +233,20 @@ func (b *backupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
+		log.Debug("Backup has in progress status from prior reconcile, marking it as failed")
+		failedCopy := original.DeepCopy()
+		failedCopy.Status.Phase = velerov1api.BackupPhaseFailed
+		failedCopy.Status.FailureReason = "Backup from previous reconcile still in progress"
```
We may want to suggest an APIServer failure here.
"Backup from previous reconcile still in progress. The API Server may have been down."

sseago commented on pkg/controller/backup_controller.go:
```diff
@@ -249,7 +263,6 @@ func (b *backupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		request.Status.Phase = velerov1api.BackupPhaseInProgress
 		request.Status.StartTimestamp = &metav1.Time{Time: b.clock.Now()}
 	}
-
```
To minimize change from upstream since this is in our fork, lets not include whitespace changes like this.

sseago commented on pkg/controller/restore_controller.go:
```diff
@@ -197,29 +197,40 @@ func (r *restoreReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+		log.Debug("Restore has in progress status from prior reconcile, marking it as failed")
+		failedCopy := original.DeepCopy()
+		failedCopy.Status.Phase = api.RestorePhaseFailed
+		failedCopy.Status.FailureReason = "Restore from previous reconcile still in progress"
```
We may want to suggest an APIServer failure here.
"Restore from previous reconcile still in progress. The API Server may have been down."

sseago commented on pkg/controller/restore_controller.go:
```diff
@@ -162,8 +162,8 @@ func (r *restoreReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 
-	restore := &api.Restore{}
-	err := r.kbClient.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, restore)
+	original := &api.Restore{}
```
Is the variable name change necessary here? This makes the diff larger and increases the possibility of rebase conflicts, since we're carrying this commit in our fork.

kaovilai replied:
I did this so restore_controller and backup_controller has the same var name pattern that's all.

kaovilai replied:
Not necessary I agree, just make the logic more pastable across both.

sseago replied:
@kaovilai I figured that's why you did that. I reverted that part here in a later commit since we're carrying the commit in our fork right now, and it removed a lot of lines from the diff, making rebase conflicts less likely. Does that seem reasonable to you?

kaovilai replied:
yes. That's reasonable

sseago reviewed (changes requested):
A couple of minor comments on error messages and the change of variable name. The bigger issue is we need the change for the backup finalizer controller as well (no restore finalizer controller in Velero 1.12, so that's not a concern here).

sseago commented:
> If patch fails on finalizer controller we wanna retry patch as fail? It's kinda difficult there since it's currently using defer func.
> 
> I would have to break the patch call out of defer to return reconciler err on patch fail. Is that ok with you?

Hmm. I think we need to do this in some way. If we can't do it with defer, then we'll need to eliminate the defer call and include this with all return statements.

sseago committed:
cleanup to minimize diff from upstream

Signed-off-by: Scott Seago <sseago@redhat.com>

sseago committed:
error message update

Signed-off-by: Scott Seago <sseago@redhat.com>

sseago committed:
requeue on finalize status update.

Unlike the InProgress transition, there's no need to fail here,
since the Finalize steps can be repeated.

sseago committed:
Only run patch once for all backup finalizer return scenarios

Signed-off-by: Scott Seago <sseago@redhat.com>

mateusoliveira43 commented on pkg/controller/backup_finalizer_controller.go:
```diff
@@ -106,15 +106,21 @@ func (r *backupFinalizerReconciler) Reconcile(ctx context.Context, req ctrl.Requ
-			return
+		if !patched {
+			if err := r.client.Patch(ctx, backup, kbclient.MergeFrom(original)); err != nil {
+				log.WithError(err).Error("Error updating backup")
```
should we update `patchErr` here and not return?

mateusoliveira43 replied:
does not make sense, it would be true and would not trigger following if

resolving

mateusoliveira43 commented on pkg/controller/backup_finalizer_controller.go:
we are not doing the same thing for restores, because it does not exist a `restore_finalizer_controller`, right?

sseago replied:
That's right. Restore finalizer controller is in Velero 1.14, but not in 1.12

kaovilai commented:
lgtm

weshayutin commented:
ok.. testing update: 

w/o the patch, the backup stayed in progress.  While updating the dpa the velero server was restarted.

backup: westest-vsphere-apidown-1

```
status:
  completionTimestamp: "2024-06-17T17:25:58Z"
  expiration: "2024-07-17T17:12:21Z"
  failureReason: found a backup with status "InProgress" during the server starting,
    mark it as "Failed"
```
1.3.0 have to change the csv to get the test image on:

```
oc get csv oadp-operator.v1.3.0 -o yaml | grep sseago
                  value: quay.io/sseago/velero:1.3-requeue
```
Once patched, I initiated a second backup and then took down the api server for roughly 1 minute:

backup: westest-vsphere-apidown-2

```
status:
  expiration: "2024-07-17T17:38:01Z"
  failureReason: Backup from previous reconcile still in progress. The API Server
    may have been down.
  formatVersion: 1.1.0
  phase: Failed
  startTimestamp: "2024-06-17T17:38:01Z"
  version: 1
```

weshayutin reviewed (approved):
/LGTM

sseago reviewed (approved):
Reviewing to remove my prior "changes requested", but don't count it as an ack, since my own changes are in here as well.

mateusoliveira43 commented:
Lets trying modifying here https://github.com/openshift/velero/blob/oadp-1.3/Makefile.prow#L61 from latest to v0.17.5 to see if CI passes?

kaovilai commented:
Just noting that this may never upstream based on comments at https://github.com/vmware-tanzu/velero/pull/7863#discussion_r1636304591

sseago commented on Makefile.prow:
```diff
@@ -58,7 +58,7 @@ envtest: $(GOBIN)/setup-envtest
 $(GOBIN)/setup-envtest:
 	@echo Installing envtest tools
-	GOFLAGS= go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
+	GOFLAGS= go install sigs.k8s.io/controller-runtime/tools/setup-envtest@v0.0.0-20230216140739-c98506dc3b8e
```
This really needs to be in a separate PR. Since this PR is for velero changes that aren't (yet?) upstream, if they ever did make it upstream, we'd be dropping this commit, and then we'd lose the prow fix.

kaovilai replied:
so merge that before this?

sseago replied:
I think so, yes. Also, won't we need that same fix elsewhere (although maybe it's already on 1.4 and konveyor-dev). Just trying to keep carried patches from upstream code limited to just those changes.

kaovilai replied:
[will drop commit once test passes and open new pr](https://github.com/openshift/velero/pull/320)

kaovilai replied:
https://github.com/openshift/velero/pull/320

kaovilai replied:
reverted.

kaovilai commented:
retest after https://github.com/openshift/velero/pull/320

kaovilai commented:
/retest

kaovilai commented:
follow on bugfix: https://github.com/openshift/velero/pull/324
